lib/commons/dom/get-text-element-stack.js

```diff
@@ -22,51 +21,7 @@ function getTextElementStack(node) {
     return [];
   }
 
-  // for code blocks that use syntax highlighting, you can get a ton of client
-  // rects (See https://github.com/dequelabs/axe-core/issues/1985). they use
-  // a mixture of text nodes and other nodes (which will contain their own text
-  // nodes), but all we care about is checking the direct text nodes as the
-  // other nodes will have their own client rects checked. doing this speeds up
-  // color contrast significantly for large syntax highlighted code blocks
-  const nodeRect = vNode.boundingClientRect;
-  const clientRects = [];
-  Array.from(node.childNodes).forEach(elm => {
-    if (elm.nodeType === 3 && sanitize(elm.textContent) !== '') {
-      const range = document.createRange();
-      range.selectNodeContents(elm);
-      const rects = Array.from(range.getClientRects());
-      /**
-       * if any text rect is larger than the bounds of the parent,
-       * or goes outside of the bounds of the parent, we need to use
-       * the parent rect so we stay within the bounds of the element.
-       *
-       * since we use the midpoint of the element when determining
-       * the rect stack we will also use the midpoint of the text rect
-       * to determine out of bounds.
-       *
-       * @see https://github.com/dequelabs/axe-core/issues/2178
-       * @see https://github.com/dequelabs/axe-core/issues/2483
-       * @see https://github.com/dequelabs/axe-core/issues/2681
-       */
-      const outsideRectBounds = rects.some(rect => {
-        const centerPoint = getRectCenter(rect);
-        return !isPointInRect(centerPoint, nodeRect);
-      });
-      if (outsideRectBounds) {
-        return;
-      }
-
-      for (const rect of rects) {
-        // filter out 0 width and height rects (newline characters)
-        // ie11 has newline characters return 0.00998, so we'll say if the
-        // line is < 1 it shouldn't be counted
-        if (rect.width >= 1 && rect.height >= 1) {
-          clientRects.push(rect);
-        }
-      }
-    }
-  });
-
+  const clientRects = getVisibleChildTextRects(node);
   if (!clientRects.length) {
     return [getElementStack(node)];
```

Comment:
I got a little confused at this point. If there's no text visible text, why doesn't this go into incomplete straight away, or return an empty array?

Reply:
The main reason is that it will handle cases where the text node is larger than the bounds of the element. That way we will use the node bounds for our calculations rather than the next node bounds. (All [the issues listed here](https://github.com/dequelabs/axe-core/pull/3703/files#diff-a443be069bc48f7824cb773157079a125dd30cd512172620ad750df4817199dbR60-R71))

```js
<div width="10px">Hello world</div>
```

Reply:
I see. It's not intuitive to me that clientRects would be empty if the child texts are larger than their parent. It seems to me like maybe we should then deal with that. That's still text that can be on an element with bad contrast. For the sake of not introducing further scope creep I'd be okay with punting that to an issue.


I think if we don't want to handle child texts that are larger than their parent in this PR, then maybe `getVisibleChildTextRects` should return the bounding box of the parent, so that it can also be used here:
https://github.com/dequelabs/axe-core/pull/3703/files#diff-eaabb887841937caf303235ad7ee968907755aa5ff002929ab03096cf848b2e5R78

If I read this right, if the text is larger than the element, it will never get reported as `elmPartiallyObscured`. Since if you pass an empty `clientRect` into `fullyEncompasses` it always returns `true`. If `clientRect` had the bounding box of the element that wouldn't be a problem.

Reply:
Great suggestion